{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sun Oct 20 15:55:58 2024       \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |\n",
      "|-----------------------------------------+------------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                        |               MIG M. |\n",
      "|=========================================+========================+======================|\n",
      "|   0  NVIDIA GeForce RTX 4090        Off |   00000000:4B:00.0 Off |                  Off |\n",
      "|  0%   31C    P8             20W /  450W |     277MiB /  24564MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "|   1  NVIDIA GeForce RTX 4090        Off |   00000000:B1:00.0 Off |                  Off |\n",
      "|  0%   31C    P8             24W /  450W |      27MiB /  24564MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "                                                                                         \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                              |\n",
      "|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |\n",
      "|        ID   ID                                                               Usage      |\n",
      "|=========================================================================================|\n",
      "|    0   N/A  N/A      4471      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "|    0   N/A  N/A     29175      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "|    0   N/A  N/A   3029231      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "|    0   N/A  N/A   3106382      G   ...bo/anaconda3/envs/mujoco/bin/python        246MiB |\n",
      "|    1   N/A  N/A      4471      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "|    1   N/A  N/A     29175      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "|    1   N/A  N/A   3029231      G   /usr/lib/xorg/Xorg                              4MiB |\n",
      "+-----------------------------------------------------------------------------------------+\n",
      "env: MUJOCO_GL=egl\n",
      "Checking that the dm_control installation succeeded...\n",
      "Installed dm_control 1.0.22\n"
     ]
    }
   ],
   "source": [
    "#@title Run to install MuJoCo and `dm_control`\n",
    "import distutils.util\n",
    "import os\n",
    "import subprocess\n",
    "if subprocess.run('nvidia-smi').returncode:\n",
    "  raise RuntimeError(\n",
    "      'Cannot communicate with GPU. '\n",
    "      'Make sure you are using a GPU Colab runtime. '\n",
    "      'Go to the Runtime menu and select Choose runtime type.')\n",
    "\n",
    "# Add an ICD config so that glvnd can pick up the Nvidia EGL driver.\n",
    "# This is usually installed as part of an Nvidia driver package, but the Colab\n",
    "# kernel doesn't install its driver via APT, and as a result the ICD is missing.\n",
    "# (https://github.com/NVIDIA/libglvnd/blob/master/src/EGL/icd_enumeration.md)\n",
    "NVIDIA_ICD_CONFIG_PATH = '/usr/share/glvnd/egl_vendor.d/10_nvidia.json'\n",
    "if not os.path.exists(NVIDIA_ICD_CONFIG_PATH):\n",
    "  with open(NVIDIA_ICD_CONFIG_PATH, 'w') as f:\n",
    "    f.write(\"\"\"{\n",
    "    \"file_format_version\" : \"1.0.0\",\n",
    "    \"ICD\" : {\n",
    "        \"library_path\" : \"libEGL_nvidia.so.0\"\n",
    "    }\n",
    "}\n",
    "\"\"\")\n",
    "\n",
    "# print('Installing dm_control...')\n",
    "# !pip install -q dm_control>=1.0.18\n",
    "\n",
    "# Configure dm_control to use the EGL rendering backend (requires GPU)\n",
    "%env MUJOCO_GL=egl\n",
    "\n",
    "print('Checking that the dm_control installation succeeded...')\n",
    "try:\n",
    "  from dm_control import suite\n",
    "  env = suite.load('cartpole', 'swingup')\n",
    "  pixels = env.physics.render()\n",
    "except Exception as e:\n",
    "  raise e from RuntimeError(\n",
    "      'Something went wrong during installation. Check the shell output above '\n",
    "      'for more information.\\n'\n",
    "      'If using a hosted Colab runtime, make sure you enable GPU acceleration '\n",
    "      'by going to the Runtime menu and selecting \"Choose runtime type\".')\n",
    "else:\n",
    "  del pixels, suite\n",
    "\n",
    "!echo Installed dm_control $(pip show dm_control | grep -Po \"(?<=Version: ).+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Other imports and helper functions\n",
    "\n",
    "# General\n",
    "import copy\n",
    "import os\n",
    "import itertools\n",
    "from IPython.display import clear_output\n",
    "import numpy as np\n",
    "\n",
    "# Graphics-related\n",
    "import matplotlib\n",
    "import matplotlib.animation as animation\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import HTML\n",
    "import PIL.Image\n",
    "# Internal loading of video libraries.\n",
    "\n",
    "# Use svg backend for figure rendering\n",
    "%config InlineBackend.figure_format = 'svg'\n",
    "\n",
    "# Font sizes\n",
    "SMALL_SIZE = 8\n",
    "MEDIUM_SIZE = 10\n",
    "BIGGER_SIZE = 12\n",
    "plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "# Inline video helper function\n",
    "if os.environ.get('COLAB_NOTEBOOK_TEST', False):\n",
    "  # We skip video generation during tests, as it is quite expensive.\n",
    "  display_video = lambda *args, **kwargs: None\n",
    "else:\n",
    "  def display_video(frames, framerate=30):\n",
    "    height, width, _ = frames[0].shape\n",
    "    dpi = 70\n",
    "\n",
    "    orig_backend = matplotlib.get_backend()\n",
    "    matplotlib.use('Agg')  # Switch to headless 'Agg' to inhibit figure rendering.\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(width / dpi, height / dpi), dpi=dpi)\n",
    "    # fig, ax = plt.subplots(1, 1, figsize=(scaled_width / dpi, scaled_height / dpi), dpi=dpi)\n",
    "    matplotlib.use(orig_backend)  # Switch back to the original backend.\n",
    "    ax.set_axis_off()\n",
    "    ax.set_aspect('equal')\n",
    "    ax.set_position([0, 0, 1, 1])\n",
    "    im = ax.imshow(frames[0])\n",
    "    def update(frame):\n",
    "      im.set_data(frame)\n",
    "      return [im]\n",
    "    interval = 1000/framerate\n",
    "    anim = animation.FuncAnimation(fig=fig, func=update, frames=frames,\n",
    "                                   interval=interval, blit=True, repeat=False)\n",
    "    return HTML(anim.to_html5_video())\n",
    "\n",
    "# Seed numpy's global RNG so that cell outputs are deterministic. We also try to\n",
    "# use RandomState instances that are local to a single cell wherever possible.\n",
    "np.random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the \"env\" folder to Python's module search path\n",
    "env_path = \"/home/zhangzhibo/CyberSpine\"\n",
    "if env_path not in sys.path:\n",
    "    sys.path.append(env_path)\n",
    "\n",
    "# print(sys.path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-10-20 15:15:13,983\tINFO worker.py:1783 -- Started a local Ray instance.\n",
      "2024-10-20 15:15:14,040\tINFO packaging.py:530 -- Creating a file package for local directory '/home/zhangzhibo/CyberSpine'.\n",
      "2024-10-20 15:15:14,103\tINFO packaging.py:358 -- Pushing file package 'gcs://_ray_pkg_361d8c0cd4ec96a7.zip' (5.60MiB) to Ray cluster...\n",
      "2024-10-20 15:15:14,216\tINFO packaging.py:371 -- Successfully pushed file package 'gcs://_ray_pkg_361d8c0cd4ec96a7.zip'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/home/zhangzhibo/CyberSpine', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/thirdparty_files', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/workers', '/tmp/ray/session_2024-10-20_15-15-11_472433_3165265/runtime_resources/py_modules_files/_ray_pkg_361d8c0cd4ec96a7', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python311.zip', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/lib-dynload', '/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages']\n"
     ]
    }
   ],
   "source": [
    "import ray\n",
    "\n",
    "ray.shutdown()\n",
    "ray.init(runtime_env={\"py_modules\": [\"/home/zhangzhibo/CyberSpine\"]})\n",
    "\n",
    "@ray.remote\n",
    "def check_path():\n",
    "    import sys\n",
    "    return sys.path\n",
    "\n",
    "print(ray.get(check_path.remote()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from env import mice_env\n",
    "import shimmy\n",
    "from dm_control import suite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gym_mice_env = shimmy.DmControlCompatibilityV0(mice_env.rodent_maze_forage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gymnasium.spaces.dict.Dict"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(gym_mice_env.observation_space)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ray import tune\n",
    "from shimmy.dm_control_compatibility import DmControlCompatibilityV0\n",
    "from env import mice_env\n",
    "\n",
    "def create_custom_env(cfg):\n",
    "    return DmControlCompatibilityV0(mice_env.rodent_maze_forage())\n",
    "\n",
    "# Register the environment with Ray\n",
    "tune.register_env(\"MiceEnv-v0\", lambda cfg: create_custom_env(cfg))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/spaces/box.py:130: UserWarning: \u001b[33mWARN: Box bound precision lowered by casting to float32\u001b[0m\n",
      "  gym.logger.warn(f\"Box bound precision lowered by casting to {self.dtype}\")\n",
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:164: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "  logger.warn(\n",
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:188: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "  logger.warn(f\"{pre} is not within the observation space.\")\n"
     ]
    }
   ],
   "source": [
    "from ray import tune\n",
    "from ray.rllib.algorithms.dreamerv3.dreamerv3 import DreamerV3Config\n",
    "\n",
    "\n",
    "def _env_creator(ctx):\n",
    "    from env import mice_env\n",
    "    import gymnasium as gym\n",
    "    from supersuit.generic_wrappers import resize_v1\n",
    "    from ray.rllib.algorithms.dreamerv3.utils.env_runner import NormalizedImageEnv\n",
    "\n",
    "    return NormalizedImageEnv(\n",
    "        resize_v1(  # resize to 64x64 and normalize images\n",
    "            gym.make(mice_env.rodent_maze_forage(), audio_on=False), x_size=64, y_size=64\n",
    "        )\n",
    "    )\n",
    "\n",
    "\n",
    "# Register the FlappyBird-rgb-v0 env including necessary wrappers via the\n",
    "# `tune.register_env()` API.\n",
    "tune.register_env(\"CyberMice\", _env_creator)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"tuneStatus\">\n",
       "  <div style=\"display: flex;flex-direction: row\">\n",
       "    <div style=\"display: flex;flex-direction: column;\">\n",
       "      <h3>Tune Status</h3>\n",
       "      <table>\n",
       "<tbody>\n",
       "<tr><td>Current time:</td><td>2024-10-20 15:54:15</td></tr>\n",
       "<tr><td>Running for: </td><td>00:00:08.65        </td></tr>\n",
       "<tr><td>Memory:      </td><td>30.1/503.5 GiB     </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "    <div class=\"systemInfo\">\n",
       "      <h3>System Info</h3>\n",
       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 0/144 CPUs, 0/2 GPUs (0.0/1.0 accelerator_type:G)\n",
       "    </div>\n",
       "    <div class=\"vDivider\"></div>\n",
       "<div class=\"messages\">\n",
       "  <h3>Messages</h3>\n",
       "  \n",
       "  \n",
       "  Number of errored trials: 1<br><table>\n",
       "<thead>\n",
       "<tr><th>Trial name                     </th><th style=\"text-align: right;\">  # failures</th><th>error file                                                                                                                                                                                      </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>DreamerV3_CyberMice_7b33c_00000</td><td style=\"text-align: right;\">           1</td><td>/tmp/ray/session_2024-10-20_15-54-03_312973_3204871/artifacts/2024-10-20_15-54-06/DreamerV3_2024-10-20_15-54-03/driver_artifacts/DreamerV3_CyberMice_7b33c_00000_0_2024-10-20_15-54-06/error.txt</td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>\n",
       "<style>\n",
       ".messages {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  padding-left: 1em;\n",
       "  overflow-y: auto;\n",
       "}\n",
       ".messages h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n",
       "\n",
       "  </div>\n",
       "  <div class=\"hDivider\"></div>\n",
       "  <div class=\"trialStatus\">\n",
       "    <h3>Trial Status</h3>\n",
       "    <table>\n",
       "<thead>\n",
       "<tr><th>Trial name                     </th><th>status  </th><th>loc  </th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td>DreamerV3_CyberMice_7b33c_00000</td><td>ERROR   </td><td>     </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "  </div>\n",
       "</div>\n",
       "<style>\n",
       ".tuneStatus {\n",
       "  color: var(--jp-ui-font-color1);\n",
       "}\n",
       ".tuneStatus .systemInfo {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus td {\n",
       "  white-space: nowrap;\n",
       "}\n",
       ".tuneStatus .trialStatus {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       ".tuneStatus h3 {\n",
       "  font-weight: bold;\n",
       "}\n",
       ".tuneStatus .hDivider {\n",
       "  border-bottom-width: var(--jp-border-width);\n",
       "  border-bottom-color: var(--jp-border-color0);\n",
       "  border-bottom-style: solid;\n",
       "}\n",
       ".tuneStatus .vDivider {\n",
       "  border-left-width: var(--jp-border-width);\n",
       "  border-left-color: var(--jp-border-color0);\n",
       "  border-left-style: solid;\n",
       "  margin: 0.5em 1em 0.5em 1em;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36m(pid=3218493)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/spaces/box.py:130: UserWarning: \u001b[33mWARN: Box bound precision lowered by casting to float32\u001b[0m\n",
      "\u001b[36m(pid=3218493)\u001b[0m   gym.logger.warn(f\"Box bound precision lowered by casting to {self.dtype}\")\n",
      "\u001b[36m(pid=3218493)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:164: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "\u001b[36m(pid=3218493)\u001b[0m   logger.warn(\n",
      "\u001b[36m(pid=3218493)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:188: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "\u001b[36m(pid=3218493)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\n",
      "2024-10-20 15:54:15,088\tERROR tune_controller.py:1331 -- Trial task failed for trial DreamerV3_CyberMice_7b33c_00000\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/air/execution/_internal/event_manager.py\", line 110, in resolve_future\n",
      "    result = ray.get(future)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/auto_init_hook.py\", line 21, in auto_init_wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/client_mode_hook.py\", line 103, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/worker.py\", line 2661, in get\n",
      "    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)\n",
      "                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/worker.py\", line 873, in get_objects\n",
      "    raise value\n",
      "ray.exceptions.ActorDiedError: The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3.__init__()\u001b[39m (pid=3218493, ip=192.168.10.49, actor_id=22aa4b3ba25057a976a8390c01000000, repr=DreamerV3)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py\", line 573, in __init__\n",
      "    super().__init__(\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/trainable/trainable.py\", line 158, in __init__\n",
      "    self.setup(copy.deepcopy(self.config))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/dreamerv3.py\", line 492, in setup\n",
      "    super().setup(config)\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py\", line 659, in setup\n",
      "    self.env_runner_group = EnvRunnerGroup(\n",
      "                            ^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 189, in __init__\n",
      "    self._setup(\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 280, in _setup\n",
      "    self._local_env_runner = self._make_worker(\n",
      "                             ^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 1173, in _make_worker\n",
      "    worker = cls(\n",
      "             ^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "    self.env = gym.vector.make(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "                                                        ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                 ^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "    env = gym.envs.registration.make(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "    env = env_creator(**env_spec_kwargs)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3204871/342441018.py\", line 8, in _env_creator\n",
      "ModuleNotFoundError: No module named 'supersuit'\n",
      "2024-10-20 15:54:15,096\tINFO tune.py:1009 -- Wrote the latest version of all result files and experiment state to '/home/zhangzhibo/ray_results/DreamerV3_2024-10-20_15-54-03' in 0.0045s.\n",
      "2024-10-20 15:54:15,100\tERROR tune.py:1037 -- Trials did not complete: [DreamerV3_CyberMice_7b33c_00000]\n",
      "2024-10-20 15:54:15,101\tINFO tune.py:1041 -- Total run time: 8.73 seconds (8.65 seconds for the tuning loop).\n",
      "2024-10-20 15:54:15,105\tWARNING experiment_analysis.py:180 -- Failed to fetch metrics for 1 trial(s):\n",
      "- DreamerV3_CyberMice_7b33c_00000: FileNotFoundError('Could not fetch metrics for DreamerV3_CyberMice_7b33c_00000: both result.json and progress.csv were not found at /home/zhangzhibo/ray_results/DreamerV3_2024-10-20_15-54-03/DreamerV3_CyberMice_7b33c_00000_0_2024-10-20_15-54-06')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m Exception raised in creation task: The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3.__init__()\u001b[39m (pid=3218493, ip=192.168.10.49, actor_id=22aa4b3ba25057a976a8390c01000000, repr=DreamerV3)\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py\", line 573, in __init__\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     super().__init__(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/trainable/trainable.py\", line 158, in __init__\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self.setup(copy.deepcopy(self.config))\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/dreamerv3.py\", line 492, in setup\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     super().setup(config)\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py\", line 659, in setup\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self.env_runner_group = EnvRunnerGroup(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                             ^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 189, in __init__\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self._setup(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 280, in _setup\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self._local_env_runner = self._make_worker(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                              ^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py\", line 1173, in _make_worker\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     worker = cls(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m              ^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self.env = gym.vector.make(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                ^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                                                         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self.envs = [env_fn() for env_fn in env_fns]\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     self.envs = [env_fn() for env_fn in env_fns]\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m                  ^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     env = gym.envs.registration.make(\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m     env = env_creator(**env_spec_kwargs)\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m   File \"/tmp/ipykernel_3204871/342441018.py\", line 8, in _env_creator\n",
      "\u001b[36m(DreamerV3 pid=3218493)\u001b[0m ModuleNotFoundError: No module named 'supersuit'\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Define the `config` variable to use for training.\n",
    "config = (\n",
    "    DreamerV3Config()\n",
    "    # set the env to the pre-registered string\n",
    "    .environment(\"CyberMice\")\n",
    "    # play around with the insanely high number of hyperparameters for DreamerV3 ;) \n",
    "    .training(\n",
    "        model_size=\"S\",\n",
    "        training_ratio=1024,\n",
    "    )\n",
    ")\n",
    "\n",
    "# Run the tuner job.\n",
    "results = tune.Tuner(trainable=\"DreamerV3\", param_space=config).fit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug Record\n",
    "\n",
    "24/10/15 Debug:\n",
    "1. parser.parse_args() 的问题在于 Jupyter Notebook 本身不通过命令行运行，因此无法传递参数给 argparse。\n",
    "2. 如果你想在 .ipynb 文件中传递参数，可以手动设置参数，而不是从命令行解析。\n",
    "\n",
    "24/10/16 Debug:\n",
    "1. RLlib 不接受用function定义的环境作为.environment的输入，\n",
    "2. 必须将maze_forazing()改为gym环境或者自定义为class变量输入给env\n",
    "3. 重点考虑手动定义class的方法\n",
    "\n",
    "24/10/18 Debug:\n",
    "1. 仔细看RLlib 定义custom env的方法 https://docs.ray.io/en/latest/rllib/rllib-env.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py:557: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "`UnifiedLogger` will be removed in Ray 2.7.\n",
      "  return UnifiedLogger(config, logdir, loggers=None)\n",
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/logger/unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `JsonLogger interface is deprecated in favor of the `ray.tune.json.JsonLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/logger/unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `CSVLogger interface is deprecated in favor of the `ray.tune.csv.CSVLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/logger/unified.py:53: RayDeprecationWarning: This API is deprecated and may be removed in future Ray releases. You could suppress this warning by setting env variable PYTHONWARNINGS=\"ignore::DeprecationWarning\"\n",
      "The `TBXLogger interface is deprecated in favor of the `ray.tune.tensorboardx.TBXLoggerCallback` interface and will be removed in Ray 2.7.\n",
      "  self._loggers.append(cls(self.config, self.logdir, self.trial))\n",
      "2024-10-20 15:56:46,528\tINFO worker.py:1783 -- Started a local Ray instance.\n",
      "\u001b[36m(pid=3232620)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/spaces/box.py:130: UserWarning: \u001b[33mWARN: Box bound precision lowered by casting to float32\u001b[0m\n",
      "\u001b[36m(pid=3232620)\u001b[0m   gym.logger.warn(f\"Box bound precision lowered by casting to {self.dtype}\")\n",
      "\u001b[36m(pid=3232620)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:164: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method was expecting numpy array dtype to be float32, actual type: float64\u001b[0m\n",
      "\u001b[36m(pid=3232620)\u001b[0m   logger.warn(\n",
      "\u001b[36m(pid=3232620)\u001b[0m /home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:188: UserWarning: \u001b[33mWARN: The obs returned by the `reset()` method is not within the observation space.\u001b[0m\n",
      "\u001b[36m(pid=3232620)\u001b[0m   logger.warn(f\"{pre} is not within the observation space.\")\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m Exception raised in creation task: The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232620, ip=192.168.10.49, actor_id=7fd65cfa5ef2e04c0a1f82a701000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x7c875edd1290>)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     self.env = gym.vector.make(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m                ^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m                                                         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     self.envs = [env_fn() for env_fn in env_fns]\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     self.envs = [env_fn() for env_fn in env_fns]\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m                  ^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     env = gym.envs.registration.make(\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     env = env_creator(**env_spec_kwargs)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m   File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m     assert isinstance(id, str)\n",
      "\u001b[36m(DreamerV3EnvRunner pid=3232620)\u001b[0m AssertionError\n",
      "2024-10-20 15:56:56,578\tERROR actor_manager.py:523 -- Ray error, taking actor 1 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232619, ip=192.168.10.49, actor_id=313ec6d69b151e690448e86901000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x79b51edd0150>)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "    self.env = gym.vector.make(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "                                                        ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                 ^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "    env = gym.envs.registration.make(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "    env = env_creator(**env_spec_kwargs)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n",
      "    assert isinstance(id, str)\n",
      "AssertionError\n",
      "2024-10-20 15:56:56,580\tERROR actor_manager.py:523 -- Ray error, taking actor 2 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232621, ip=192.168.10.49, actor_id=cd6c970bdbea8dad1ec80ae201000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x787bbf3efa90>)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "    self.env = gym.vector.make(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "                                                        ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                 ^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "    env = gym.envs.registration.make(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "    env = env_creator(**env_spec_kwargs)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n",
      "    assert isinstance(id, str)\n",
      "AssertionError\n",
      "2024-10-20 15:56:56,582\tERROR actor_manager.py:523 -- Ray error, taking actor 3 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232620, ip=192.168.10.49, actor_id=7fd65cfa5ef2e04c0a1f82a701000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x7c875edd1290>)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "    self.env = gym.vector.make(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "                                                        ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                 ^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "    env = gym.envs.registration.make(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "    env = env_creator(**env_spec_kwargs)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n",
      "    assert isinstance(id, str)\n",
      "AssertionError\n",
      "2024-10-20 15:56:56,586\tERROR actor_manager.py:523 -- Ray error, taking actor 4 out of service. The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232622, ip=192.168.10.49, actor_id=dc2ed011b1c3baa713ea226101000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x70279b400890>)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n",
      "    self.env = gym.vector.make(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n",
      "    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n",
      "                                                        ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n",
      "    self.envs = [env_fn() for env_fn in env_fns]\n",
      "                 ^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n",
      "    env = gym.envs.registration.make(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n",
      "    env = env_creator(**env_spec_kwargs)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n",
      "  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n",
      "    assert isinstance(id, str)\n",
      "AssertionError\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mActorDiedError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py:189\u001b[0m, in \u001b[0;36mEnvRunnerGroup.__init__\u001b[0;34m(self, env_creator, validate_env, default_policy_class, config, num_env_runners, local_env_runner, logdir, _setup, tune_trial_id, num_workers, local_worker)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 189\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_setup\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    190\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    191\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    192\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    193\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlocal_env_runner\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlocal_env_runner\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    194\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    195\u001b[0m \u001b[38;5;66;03m# EnvRunnerGroup creation possibly fails, if some (remote) workers cannot\u001b[39;00m\n\u001b[1;32m    196\u001b[0m \u001b[38;5;66;03m# be initialized properly (due to some errors in the EnvRunners's\u001b[39;00m\n\u001b[1;32m    197\u001b[0m \u001b[38;5;66;03m# constructor).\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py:259\u001b[0m, in \u001b[0;36mEnvRunnerGroup._setup\u001b[0;34m(self, validate_env, config, num_env_runners, local_env_runner)\u001b[0m\n\u001b[1;32m    258\u001b[0m \u001b[38;5;66;03m# Create a number of @ray.remote workers.\u001b[39;00m\n\u001b[0;32m--> 259\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd_workers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    260\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    261\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate_env_runners_after_construction\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    262\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    264\u001b[0m \u001b[38;5;66;03m# If num_env_runners > 0 and we don't have an env on the local worker,\u001b[39;00m\n\u001b[1;32m    265\u001b[0m \u001b[38;5;66;03m# get the observation- and action spaces for each policy from\u001b[39;00m\n\u001b[1;32m    266\u001b[0m \u001b[38;5;66;03m# the first remote worker (which does have an env).\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py:792\u001b[0m, in \u001b[0;36mEnvRunnerGroup.add_workers\u001b[0;34m(self, num_workers, validate)\u001b[0m\n\u001b[1;32m    791\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m result\u001b[38;5;241m.\u001b[39mok:\n\u001b[0;32m--> 792\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m result\u001b[38;5;241m.\u001b[39mget()\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/utils/actor_manager.py:500\u001b[0m, in \u001b[0;36mFaultTolerantActorManager._fetch_result\u001b[0;34m(self, remote_actor_ids, remote_calls, tags, timeout_seconds, return_obj_refs, mark_healthy)\u001b[0m\n\u001b[1;32m    499\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 500\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mray\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mready\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    501\u001b[0m     remote_results\u001b[38;5;241m.\u001b[39madd_result(actor_id, ResultOrError(result\u001b[38;5;241m=\u001b[39mresult), tag)\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/auto_init_hook.py:21\u001b[0m, in \u001b[0;36mwrap_auto_init.<locals>.auto_init_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     20\u001b[0m auto_init_ray()\n\u001b[0;32m---> 21\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/client_mode_hook.py:103\u001b[0m, in \u001b[0;36mclient_mode_hook.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(ray, func\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m)(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m--> 103\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/worker.py:2661\u001b[0m, in \u001b[0;36mget\u001b[0;34m(object_refs, timeout)\u001b[0m\n\u001b[1;32m   2660\u001b[0m \u001b[38;5;66;03m# TODO(ujvl): Consider how to allow user to retrieve the ready objects.\u001b[39;00m\n\u001b[0;32m-> 2661\u001b[0m values, debugger_breakpoint \u001b[38;5;241m=\u001b[39m \u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_objects\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobject_refs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2662\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, value \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(values):\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/_private/worker.py:873\u001b[0m, in \u001b[0;36mWorker.get_objects\u001b[0;34m(self, object_refs, timeout, return_exceptions)\u001b[0m\n\u001b[1;32m    872\u001b[0m             \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 873\u001b[0m                 \u001b[38;5;28;01mraise\u001b[39;00m value\n\u001b[1;32m    875\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m values, debugger_breakpoint\n",
      "\u001b[0;31mActorDiedError\u001b[0m: The actor died because of an error raised in its creation task, \u001b[36mray::DreamerV3EnvRunner.__init__()\u001b[39m (pid=3232619, ip=192.168.10.49, actor_id=313ec6d69b151e690448e86901000000, repr=<ray.rllib.algorithms.dreamerv3.utils.env_runner.DreamerV3EnvRunner object at 0x79b51edd0150>)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/utils/env_runner.py\", line 166, in __init__\n    self.env = gym.vector.make(\n               ^^^^^^^^^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 82, in make\n    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)\n                                                        ^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in __init__\n    self.envs = [env_fn() for env_fn in env_fns]\n                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/sync_vector_env.py\", line 53, in <listcomp>\n    self.envs = [env_fn() for env_fn in env_fns]\n                 ^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/vector/__init__.py\", line 60, in _make_env\n    env = gym.envs.registration.make(\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 801, in make\n    env = env_creator(**env_spec_kwargs)\n          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/tmp/ipykernel_3224087/342441018.py\", line 13, in _env_creator\n  File \"/home/zhangzhibo/anaconda3/envs/mujoco/lib/python3.11/site-packages/gymnasium/envs/registration.py\", line 737, in make\n    assert isinstance(id, str)\nAssertionError",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 53\u001b[0m\n\u001b[1;32m     18\u001b[0m config \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     19\u001b[0m     DreamerV3Config()\n\u001b[1;32m     20\u001b[0m     \u001b[38;5;66;03m# Use image observations.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     49\u001b[0m \n\u001b[1;32m     50\u001b[0m )\n\u001b[1;32m     52\u001b[0m config\u001b[38;5;241m.\u001b[39mremote_worker_envs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m---> 53\u001b[0m rllib_algo \u001b[38;5;241m=\u001b[39m \u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbuild\u001b[49m\u001b[43m(\u001b[49m\u001b[43muse_copy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm_config.py:880\u001b[0m, in \u001b[0;36mAlgorithmConfig.build\u001b[0;34m(self, env, logger_creator, use_copy)\u001b[0m\n\u001b[1;32m    877\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malgo_class, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    878\u001b[0m     algo_class \u001b[38;5;241m=\u001b[39m get_trainable_cls(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malgo_class)\n\u001b[0;32m--> 880\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgo_class\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43muse_copy\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlogger_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    883\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py:573\u001b[0m, in \u001b[0;36mAlgorithm.__init__\u001b[0;34m(self, config, env, logger_creator, **kwargs)\u001b[0m\n\u001b[1;32m    570\u001b[0m \u001b[38;5;66;03m# Evaluation EnvRunnerGroup and metrics last returned by `self.evaluate()`.\u001b[39;00m\n\u001b[1;32m    571\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39meval_env_runner_group: Optional[EnvRunnerGroup] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 573\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    574\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    575\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlogger_creator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlogger_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    576\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    577\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/tune/trainable/trainable.py:158\u001b[0m, in \u001b[0;36mTrainable.__init__\u001b[0;34m(self, config, logger_creator, storage)\u001b[0m\n\u001b[1;32m    154\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStorageContext on the TRAINABLE:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mstorage\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    156\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_open_logfiles(stdout_file, stderr_file)\n\u001b[0;32m--> 158\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msetup\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdeepcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    159\u001b[0m setup_time \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime() \u001b[38;5;241m-\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_start_time\n\u001b[1;32m    160\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m setup_time \u001b[38;5;241m>\u001b[39m SETUP_TIME_THRESHOLD:\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/dreamerv3/dreamerv3.py:492\u001b[0m, in \u001b[0;36mDreamerV3.setup\u001b[0;34m(self, config)\u001b[0m\n\u001b[1;32m    490\u001b[0m \u001b[38;5;129m@override\u001b[39m(Algorithm)\n\u001b[1;32m    491\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msetup\u001b[39m(\u001b[38;5;28mself\u001b[39m, config: AlgorithmConfig):\n\u001b[0;32m--> 492\u001b[0m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msetup\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    494\u001b[0m     \u001b[38;5;66;03m# Share RLModule between EnvRunner and single (local) Learner instance.\u001b[39;00m\n\u001b[1;32m    495\u001b[0m     \u001b[38;5;66;03m# To avoid possibly expensive weight synching step.\u001b[39;00m\n\u001b[1;32m    496\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mshare_module_between_env_runner_and_learner:\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/algorithms/algorithm.py:659\u001b[0m, in \u001b[0;36mAlgorithm.setup\u001b[0;34m(self, config)\u001b[0m\n\u001b[1;32m    656\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39moff_policy_estimation_methods \u001b[38;5;241m=\u001b[39m ope_dict\n\u001b[1;32m    658\u001b[0m \u001b[38;5;66;03m# Create a set of env runner actors via a EnvRunnerGroup.\u001b[39;00m\n\u001b[0;32m--> 659\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39menv_runner_group \u001b[38;5;241m=\u001b[39m \u001b[43mEnvRunnerGroup\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    660\u001b[0m \u001b[43m    \u001b[49m\u001b[43menv_creator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menv_creator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    661\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvalidate_env\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdefault_policy_class\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_default_policy_class\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_env_runners\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    665\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\n\u001b[1;32m    666\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    667\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minput_\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minput_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minput_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minput_\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m                \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minput_\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m!=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msampler\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m    676\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menable_rl_module_and_learner\u001b[49m\n\u001b[1;32m    677\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menable_env_runner_and_connector_v2\u001b[49m\n\u001b[1;32m    678\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    679\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnum_env_runners\u001b[49m\n\u001b[1;32m    680\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    681\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlocal_env_runner\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    682\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlogdir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlogdir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    683\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtune_trial_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrial_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    684\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    686\u001b[0m \u001b[38;5;66;03m# If an input path is available and we are on the new API stack generate\u001b[39;00m\n\u001b[1;32m    687\u001b[0m \u001b[38;5;66;03m# an `OfflineData` instance.\u001b[39;00m\n\u001b[1;32m    688\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    689\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39minput_\n\u001b[1;32m    690\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    699\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39menable_env_runner_and_connector_v2\n\u001b[1;32m    700\u001b[0m ):\n",
      "File \u001b[0;32m~/anaconda3/envs/mujoco/lib/python3.11/site-packages/ray/rllib/env/env_runner_group.py:211\u001b[0m, in \u001b[0;36mEnvRunnerGroup.__init__\u001b[0;34m(self, env_creator, validate_env, default_policy_class, config, num_env_runners, local_env_runner, logdir, _setup, tune_trial_id, num_workers, local_worker)\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m RayActorError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    199\u001b[0m     \u001b[38;5;66;03m# In case of an actor (remote worker) init failure, the remote worker\u001b[39;00m\n\u001b[1;32m    200\u001b[0m     \u001b[38;5;66;03m# may still exist and will be accessible, however, e.g. calling\u001b[39;00m\n\u001b[1;32m    201\u001b[0m     \u001b[38;5;66;03m# its `sample.remote()` would result in strange \"property not found\"\u001b[39;00m\n\u001b[1;32m    202\u001b[0m     \u001b[38;5;66;03m# errors.\u001b[39;00m\n\u001b[1;32m    203\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m e\u001b[38;5;241m.\u001b[39mactor_init_failed:\n\u001b[1;32m    204\u001b[0m         \u001b[38;5;66;03m# Raise the original error here that the EnvRunners raised\u001b[39;00m\n\u001b[1;32m    205\u001b[0m         \u001b[38;5;66;03m# during its construction process. This is to enforce transparency\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    209\u001b[0m         \u001b[38;5;66;03m# - e.args[0].args[2]: The original Exception (e.g. a ValueError due\u001b[39;00m\n\u001b[1;32m    210\u001b[0m         \u001b[38;5;66;03m# to a config mismatch) thrown inside the actor.\u001b[39;00m\n\u001b[0;32m--> 211\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39margs[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39margs[\u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# In any other case, raise the RayActorError as-is.\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    214\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import sys\n",
    "from ray.rllib.algorithms.dreamerv3.dreamerv3 import DreamerV3Config\n",
    "from ray.rllib.utils.test_utils import add_rllib_example_script_args\n",
    "\n",
    "sys.argv = ['script_name','--num-gpus','1','--num-env-runners','4']\n",
    "\n",
    "parser = add_rllib_example_script_args(\n",
    "    default_iters=1000000,\n",
    "    default_reward=800.0,\n",
    "    default_timesteps=1000000\n",
    ")\n",
    "# Use `parser` to add your own custom command line options to this script\n",
    "# and (if needed) use their values toset up `config` below.\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "\n",
    "config = (\n",
    "    DreamerV3Config()\n",
    "    # Use image observations.\n",
    "    # .environment(\n",
    "    #     env=\"MiceEnv-v0\",\n",
    "    #     env_config={\"from_pixels\": True},\n",
    "    # )\n",
    "    .environment(\"CyberMice\")\n",
    "    .learners(\n",
    "        num_learners=0 if args.num_gpus == 1 else args.num_gpus,\n",
    "        num_gpus_per_learner=1 if args.num_gpus else 0,\n",
    "    )\n",
    "    .env_runners(\n",
    "        num_env_runners=(args.num_env_runners or 0),\n",
    "        # If we use >1 GPU and increase the batch size accordingly, we should also\n",
    "        # increase the number of envs per worker.\n",
    "        num_envs_per_env_runner=4 * (args.num_gpus or 1),\n",
    "        remote_worker_envs=True,\n",
    "    )\n",
    "    .reporting(\n",
    "        metrics_num_episodes_for_smoothing=(args.num_gpus or 1),\n",
    "        report_images_and_videos=False,\n",
    "        report_dream_data=False,\n",
    "        report_individual_batch_item_stats=False,\n",
    "    )\n",
    "    # See Appendix A.\n",
    "    .training(\n",
    "        model_size=\"S\",\n",
    "        training_ratio=512,\n",
    "        batch_size_B=16 * (args.num_gpus or 1),\n",
    "    )\n",
    "\n",
    ")\n",
    "\n",
    "config.remote_worker_envs=False\n",
    "rllib_algo = config.build(use_copy=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mujoco",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
